### PR TITLE
feat: Update to WebXR

### PR DESCRIPTION
**Pull request by mister-ben**
Brings @kevleyski 's changes in #273 into main:

For devices that support WebXR, use that
Add iOS device orientation permission check on entering 360 (note ESLint workaround as this is iOS Safari specfic)
Fall back to original WebVR polyfill if navigator.xr not available after initialising WebXR polyfill
Includes fixes for projections other than equirectangular that needed updates after three.js version bump

Also:
Removes husky again, because the pre-commit lint needs a bump to at least Node 15 to accept some code in dependencies (nullish coalescing assignment), and upgrading that is a whole other package of work.
Prevents Browserstack CI from trying to run on Chromium, as the video playback will fail.

### CODE REVIEW

**Comment by kevleyski** [comment]
Great stuff, will give this a proper go later but looks good already

**Review by misteroneill — APPROVED** [review]
Thanks for porting this over, @mister-ben!

I don't know this project well, so take my review with a grain of salt, but the changes look good to me.

File: src/plugin.js [hunk 1]
```diff
@@ -762,10 +901,351 @@ void main() {
...
+  initShuttleControls() {
+    this.holodeck = new THREE.LineSegments(new BoxLineGeometry(6, 6, 6, 10, 10, 10), new THREE.MeshBasicMaterial({
+      opacity: 0,
+      transparent: true
+    }));
+    this.holodeck.geometry.translate(0, 3, 0);
+
+    this.scene.add(this.holodeck);
+
+    const controlsGeometry = new this.RoundedRectangle(2.4, 0.6, 0.05, 5.0);
```
Comment (amtins) [inline src/plugin.js:1070]: This line causes a `TypeError: this.RoundedRectangle is not a constructor`

File: src/plugin.js [hunk 2]
```diff
@@ -695,46 +789,91 @@ void main() {
...
+        }
+        window.navigator.xr.setSession = (session) => {
+          this.currentSession = session;
+          this.renderer.xr.setSession(this.currentSession);
+        };
       });
+    } else {
+      // fallback to older WebVR if WebXR Device API is not available
+      this.initVRPolyfill(displays);
+    }
```
Comment (amtins) [inline src/plugin.js:833]: The variable `displays` is set in a promise, I don't see how this could work but maybe I'm missing something.

Comment (amtins) [reply src/plugin.js:833]: This seems to lead to an NPE on the property `this.player_.vr().vrDisplay` in the following file :https://github.com/mister-ben/videojs-vr/blob/4576704f9c807b632a744607bfed8a084dedf509/src/cardboard-button.js#L42-L49

**Comment by amtins** [comment]
The progress bar circle is visible
Screenshot from 2023-12-18 19-28-12

**Comment by mister-ben** [comment]
> The progress bar circle is visible Screenshot from 2023-12-18 19-28-12

I haven't seen that... which example / which browser was this?

**Comment by amtins** [comment]
@mister-ben I have this behavior with all the examples on `Firefox` and `Chrome` on desktop. To reproduce this case, simply :
- open any `example`
- open the developer console and select the `WebRX` tab
- start playback, the media should be displayed in stereoscopic mode, if not, click on the `Enter VR` button
- click on `exit immersive` in the WebRX tab

**Comment by mister-ben** [comment]
I've removed the cardboard button in the control bar for now, and used three.js's VRButton as an overlay button. I don't love the UX, if we want the cardboard button we need to include what's going on inside the VRButton class, it's quite different from what worked before. This seems ok as an interim compromise.

This works well on Android. iOS is a pain going in and out of immersive mode. It works sometimes, other times Safari doesn't understand the screen orientation and messes up the display. When exiting, the canvas is larger than the player. I don't know yet whether the issues are in three.js, the plugin, or Safari. Or a combination.

**Comment by amtins** [comment]
@mister-ben on Android on Firefox, moving the phone to change orientation works well, but on Chrome it doesn't. Also, on VR mode, the back and settings buttons don't respond to touch, as well as when using a google cardboard the toggle play doesn't work.

**Comment by amtins** [comment]
I see that the canvas is created here [plugin.js#L783-L788](https://github.com/mister-ben/videojs-vr/blob/f32085a1dd45ce8c6efebea5aa05062e00131d69/src/plugin.js#L783-L788), and then it's deleted somehow when entering immersive mode. When leaving immersive mode, the canvas is added to the player as the last child somehow, and this is the cause of [#285 (comment)](https://github.com/videojs/videojs-vr/pull/285#issuecomment-1861377978). But I can't figure out how.

**Review by amtins — APPROVED** [review]

**Comment by WCPetersen** [comment]
This seems to break the EAC projections. Can anyone else confirm?

**Comment by kevleyski** [comment]
(do we have some good examples, happy to dive in and check on the headsets I have)On 12 Feb 2024, at 5:42 am, Will Petersen ***@***.***> wrote:﻿
This seems to break the EAC projections. Can anyone else confirm?

**Comment by WCPetersen** [comment]
Yes, there are examples for both EAC and EAC_LR in the examples/samples dirs. They should be listed when you run `npm start`. Even without the headset, they're broken in the browser. I just want to check if that is happening with everyone or if I have a bad dependency (or something like that).

File: src/plugin.js [hunk 3]
```diff
@@ -762,14 +901,356 @@ void main() {
...
+    });
+
+    function onControllerDisconnected() {
+      const index = this.userData.index;
+
+      if (self.controllers) {
+        const obj = (index === 0) ? self.controllers[0] : self.controllers[1];
+
+        if (obj) {
+          if (obj.controller) {
```
Comment (WCPetersen) [inline src/plugin.js:1221]: `obj.controller` and `obj.grip` are never set.

File: src/plugin.js [hunk 4]
```diff
@@ -762,14 +901,356 @@ void main() {
...
+    }
+
+    this.renderer.render(this.scene, this.camera);
+  }
+
+  addCardboardButton_() {
+    // Not using the cardboard button in favour of three.js's vrbutton.
+    // if (!this.player_.controlBar.getChild('CardboardButton')) {
+    //   this.player_.controlBar.addChild('CardboardButton', {});
+    // }
```
Comment (WCPetersen) [inline src/plugin.js:1253]: If the cardboard button is being removed in favor of the three.js VRButton, shouldn't all references to `addCardboardButton_()`, the whole `cardboard-button.js` file, and the `forceCardboard` option be removed? Why go halfway?

File: src/plugin.js [hunk 5]
```diff
@@ -762,14 +901,356 @@ void main() {
...
+      const grip = this.renderer.xr.getControllerGrip(i);
+
+      grip.add(controllerModelFactory.createControllerModel(grip));
+      this.scene.add(grip);
+    }
+
+    return controllers;
+  }
+
+  createText(message, height) {
```
Comment (WCPetersen) [inline src/plugin.js:958]: Does this ever get called?

**Comment by kevleyski** [comment]
Yeah controllers were working - I’ll try find some time to check if something changed On 20 Feb 2024, at 7:29 am, Will Petersen ***@***.***> wrote:﻿
@WCPetersen commented on this pull request.



In src/plugin.js:
> +      controller.userData.selectPressed = false;
+      controller.userData.index = i;
+      this.scene.add(controller);
+
+      controllers.push(controller);
+
+      const grip = this.renderer.xr.getControllerGrip(i);
+
+      grip.add(controllerModelFactory.createControllerModel(grip));
+      this.scene.add(grip);
+    }
+
+    return controllers;
+  }
+
+  createText(message, height) {

Does this ever get called?

**Comment by blysik** [comment]
As an Apple Vision Pro user, I would love to see this implemented.  Any chance it can worked on more soon?  Thanks!

**Comment by jbroberg** [comment]
> As an Apple Vision Pro user, I would love to see this implemented. Any chance it can worked on more soon? Thanks!

We are actively working on it over [here](https://github.com/metacdn/videojs-vr) - once it stabilises we can have another go at merging upstream. Any feedback is welcome - right now we are focused on improving/making reliable the entering and exiting of immersive mode on HMDs, which is not working super reliably right now.

**Comment by mister-ben** [comment]
Closing in favour of the other work

**Pull request closed without merging**